sycl/include/sycl/ext/oneapi/matrix/compile-time-query.hpp

```diff
@@ -0,0 +1,392 @@
+#include <sycl/ext/oneapi/matrix/matrix-tensorcores.hpp>
+#include <sycl/ext/oneapi/matrix/matrix-unified-utils.hpp> // for use, layout
+#include <sycl/ext/oneapi/matrix/matrix-unified.hpp>       // for joint_matrix
+
+#include <cstddef>     // for size_t
+#include <stdint.h>    // for uint32_t, int8_t
+#include <type_traits> // for enable_if
+
+using namespace sycl::ext::oneapi::experimental;
+
+namespace sycl {
+inline namespace _V1 {
+namespace ext {
+namespace oneapi {
+namespace experimental::matrix {
+
+enum class matrix_type {
+  bf16,
+  fp16,
+  tf32,
+  fp32,
+  fp64,
+  sint8,
+  sint16,
+  sint32,
+  sint64,
+  uint8,
+  uint16,
+  uint32,
+  uint64
+};
+
+struct combination {
+  size_t max_msize;
+  size_t max_nsize;
+  size_t max_ksize;
+  size_t msize;
+  size_t nsize;
+  size_t ksize;
+  matrix_type atype;
+  matrix_type btype;
+  matrix_type ctype;
+  matrix_type dtype;
+};
+
+template <architecture u, typename Ta, typename Tb, typename Tc,
+          typename Td = Tc, size_t sM = 0, size_t sN = 0, size_t sK = 0,
+          typename Enabled = void>
+struct matrix_params;
+
+template <typename Ta, typename Tb, typename Tc>
+constexpr bool is_combination_valid_amx(size_t sM, size_t sN, size_t sK) {
+  // is_same_v is a C++17 feature
+  if ((std::is_same_v<Ta, int8_t> && std::is_same_v<Tb, int8_t> &&
+       std::is_same_v<Tc, int> && sM <= 16 && sN <= 16 && sK <= 64) ||
+      (std::is_same_v<Ta, uint8_t> && std::is_same_v<Tb, uint8_t> &&
+       std::is_same_v<Tc, int> && sM <= 16 && sN <= 16 && sK <= 64) ||
+      (std::is_same_v<Ta, int8_t> && std::is_same_v<Tb, uint8_t> &&
+       std::is_same_v<Tc, int> && sM <= 16 && sN <= 16 && sK <= 64) ||
+      (std::is_same_v<Ta, uint8_t> && std::is_same_v<Tb, int8_t> &&
+       std::is_same_v<Tc, int> && sM <= 16 && sN <= 16 && sK <= 64) ||
+      // bf16
+      (std::is_same_v<Ta, unsigned short> &&
+       std::is_same_v<Tb, unsigned short> && std::is_same_v<Tc, float> &&
+       sM <= 16 && sN <= 16 && sK <= 32))
+    return true;
+  else
+    return false;
+}
+
+template <typename Ta, typename Tb, typename Tc>
+constexpr bool are_types_valid_amx() {
+  if ((std::is_same_v<Ta, int8_t> && std::is_same_v<Tb, int8_t> &&
+       std::is_same_v<Tc, int>) ||
+      (std::is_same_v<Ta, uint8_t> && std::is_same_v<Tb, uint8_t> &&
+       std::is_same_v<Tc, int>) ||
+      (std::is_same_v<Ta, int8_t> && std::is_same_v<Tb, uint8_t> &&
+       std::is_same_v<Tc, int>) ||
+      (std::is_same_v<Ta, uint8_t> && std::is_same_v<Tb, int8_t> &&
+       std::is_same_v<Tc, int>) ||
+      (std::is_same_v<Ta, unsigned short> &&
+       std::is_same_v<Tb, unsigned short> && std::is_same_v<Tc, float>))
+    return true;
+  else
+    return false;
+}
+
+// Default values query
+// Specialization for when only types are given, need to query only sizes
+template <typename Ta, typename Tb, typename Tc, typename Td>
+struct matrix_params<
+    architecture::intel_cpu_spr, Ta, Tb, Tc, Td, 0, 0, 0,
```

Comment:
Design question: can we make a dependency on experimental extension? The API of experimental extension can be changed at any time without any notice.

Reply:
joint_matrix is also experimental. The review in https://github.com/intel/llvm/pull/7964 recommended that having this is better than the names in enum tpu {amx, xmx8, xmx16 } we had before.